### PR TITLE
Trial

### PR DESCRIPTION
**Pull request by vigehi**
Design

    You should follow these [design guidelines](https://www.behance.net/gallery/19759151/Snapscan-iOs-design-and-branding?tracking_source=), including:
        Colors.
        Typography: font face, size and weight.
        Layout: composition and space between elements.

    NOTE: In these design guidelines there are several UIs that you won't need for this exercise; also, some pages are not given a design and you will create them following the design guidelines of the other pages.

Original design idea by [Gregoire Vella on Behance](https://www.behance.net/gregoirevella).

The [Creative Commons license of the design](https://creativecommons.org/licenses/by-nc/4.0/) requires that you give appropriate credit to the author. Therefore, you must do it in the README of your project.
Interactions

    Splash screen
        A simple page with the name of your app (yes, you need to choose one), and links to the sign up and log in pages.

    Sign up and log in pages
        The user should be able to register in the app with full name, email and password (all mandatory).
        The user can log into the app using email and password.
        If the user is not logged in, they can't access pages that require the user to be logged in (all the pages described below).

    Home page (catagories page)
        When the user logs in, they are presented with the categories page.
        For each category, the user can see their name, icon and the total amount of all the transactions that belongs to that category.
        When the user clicks (or taps) on a category item, the application navigates to the transactions page for that category.
        There is a button "add a new category" at the bottom that brings the user to the page to create a new category.

    Transactions page
        For a given category, the list of transactions is presented, ordered by the most recent.
        At the top of the page the user could see the total amount for the category (sum of all of the amounts of the transactions in that category).
        There is a button "add a new transaction" at the bottom that brings the user to the page to create a new transaction.
        When the user clicks on the "Back" button (<), the user navigates to the home page.

    "Add a new category" page
        The user fills out a form to create a new category, indicating their name and icon (both mandatory).
        The user click (or taps) the "Save" button to create the new category, and is taken to the home page on success.
        When the user clicks on the "Back" button (<), the user navigates to the home page.

    "Add a new transaction" page
        The user fills out a form to create a new transaction with:
            name (mandatory)
            amount (mandatory)
            categories (mandatory at least one)
        The user click (or taps) the "Save" button to create the new transation, and is taken to the transactions page for that category.
        When the user clicks on the "Back" button (<), the user navigates to the transactions page for that category.

Testing requirements

    Create unit and integration tests for all the most important components of your RoR application.

Technical requirements

    You should use Postgres as your database.
    You should use devise for authentication.
    You should validate all user input to make sure that anyone with bad intentions cannot compromise your app.
    You can use a view template engine of your choice (.erb, .slim, .haml).
    The project should be deployed and accessible online.
    Your database schema should reflect the following structure:

### CODE REVIEW

**Review by tresorsawasawa — COMMENTED** [review]
# STATUS: INVALID CODE REVIEW REQUEST 🛑 🛑

Hi @vigehi ,:wave: :wave: 
While you made a great effort in this project, unfortunately, I cannot proceed to review your code.

## Invalid Code Review Request 🛑 🛑

You have submitted a project that doesn't run. Below are an error image I am getting wen trying to run migrations:

Screenshot from 2023-01-11 22-54-37



Your Code Review Request will be marked as invalid in your Dashboard, so please submit a new one once you are ready 🙏

---
Cheers and Happy coding!👏👏👏

_Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please ping me [@tresorsawasawa](https://github.com/tresorsawasawa) when you comment so I can receive the notification or use slack with the same name._

---
_Invalid Code Review Request **Does not count** into the code reviews limit._

File: README.md [hunk 1]
```diff
@@ -1,73 +1,238 @@
...
+> Below are the key features of the application.
+
+- **[Homepage]**
+- **[Login and sign up form]**
+- **[category and transactions]**
+
+<p align="right">(<a href="#readme-top">back to top</a>)</p>
+
+
+- [Live Demo Link]()
```
Comment (andersonlebon) [inline README.md:87]: - [x] Kindly make sure you deploy your app and  put the link to the live demo here as it is said in the requirements

File: README.md [hunk 2]
```diff
@@ -1,73 +1,238 @@
...
+
+##  FutureFeaatures <a name="future"></a>
+
+> More bootstrap on the project
+
+<!-- ACKNOWLEDGEMENTS -->
+
+## 🙏 Acknowledgments <a name="acknowledgements"></a>
+
+> Give credit to everyone who inspired your codebase.
```
Comment (andersonlebon) [inline README.md:212]: - [x] Kindly it's required to give appropriate credit in the  Acknowledgments section to [Gregoire Vella on Behance](https://www.behance.net/gregoirevella) the [author of the original design](https://www.behance.net/gallery/19759151/Snapscan-iOs-design-and-branding?tracking_source=), as required by the [Creative Commons license of the design](https://creativecommons.org/licenses/by-nc/4.0/).

File: app/views/group/index.html.erb [hunk 3]
```diff
@@ -0,0 +1,21 @@
+<h3 class="mb-3">Categories</h3>
+
+<% if @groups.count > 0 %>
+  <% @groups.each do |group| %>
+    <%= link_to group.entities(group), class: "card mb-2" do %>
```
Comment (andersonlebon) [inline app/views/group/index.html.erb]: - [x] Kindly use the correct syntax to link to a single group
```html.rb
  <%= link_to group_entities_path(group), class: "card mb-2" do %>
```
image

File: app/models/entity.rb [hunk 4]
```diff
@@ -1,4 +1,9 @@
 class Entity < ApplicationRecord
-  belongs_to :user
-  has_many :groups
+  belongs_to :author, class_name: 'User', foreign_key: :author_id
+
+  has_many :entities
+  has_many :groups, through: :entities
```
Comment (andersonlebon) [inline app/models/entity.rb:5]: - [x] Kindly note that there are no foreig_keys in your tables that certify these associations you made. kindly run a migration that will certify the association you made in the above code. this is leading your app to some performance issues(infinite loops) 

image

image

Your association should match the relationship diagram below
(https://gitlab.com/microverse/guides/tse/code_review/code_review_guidelines_2_0/-/raw/master/ror/capstone/images/erd_diagram.png

**Review by andersonlebon — CHANGES_REQUESTED** [review]
## Hi @vigehi 

Good job so far!
There are some issues that you still need to work on to prepare your project for the final evaluation but you are almost there!

almost-there

## Suggested changes
_Check the comments under the review._

You can use as many of my suggestions as you want. If there is anything you would like to skip - feel free to do that. However, I strongly recommend you take them into account as they can make your code better._


Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question `@andersonlebon` so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

File: spec/models/transaction_spec.rb [hunk 5]
```diff
@@ -0,0 +1,30 @@
...
+      subject.amount = nil
+      expect(subject).to_not be_valid
+    end
+
+    it 'the amount should be numerical' do
+      subject.amount = 'a'
+      expect(subject).to_not be_valid
+    end
+  end
+end
```
Comment (andersonlebon) [inline spec/models/transaction_spec.rb]: - [ ] Kindly fix also all failing tests

image

File: README.md (hunk 1 above)
Comment (vigehi) [reply README.md:87]: done

File: README.md (hunk 2 above)
Comment (vigehi) [reply README.md:212]: done

File: app/models/entity.rb (hunk 4 above)
Comment (vigehi) [reply app/models/entity.rb:5]: done

File: spec/models/transaction_spec.rb (hunk 5 above)
Comment (vigehi) [reply spec/models/transaction_spec.rb]: done

File: app/views/group/index.html.erb (hunk 3 above)
Comment (vigehi) [reply app/views/group/index.html.erb]: done

**Pull request merged**